stubmodify/modify.go

```diff
@@ -40,41 +41,42 @@ func WriteAttributionCode(mapped, code []byte) ([]byte, error) {
 	} else if peMagicNumber == 0x20b {
 		certDirEntryOffset = optionalHeaderOffset + 144
 	} else {
-		return nil, errors.New("mapped is not in a known PE format")
+		return nil, -1, errors.New("mapped is not in a known PE format")
 	}
 
 	if len(mapped) < int(certDirEntryOffset+8) {
-		return nil, fmt.Errorf("mapped is shorter than certDirEntryOffset+8: %d", certDirEntryOffset+8)
+		return nil, -1, fmt.Errorf("mapped is shorter than certDirEntryOffset+8: %d", certDirEntryOffset+8)
 	}
 	certTableOffset := byteOrder.Uint32(mapped[certDirEntryOffset : certDirEntryOffset+4])
 	certTableSize := byteOrder.Uint32(mapped[certDirEntryOffset+4 : certDirEntryOffset+8])
 
 	if certTableOffset == 0 || certTableSize == 0 {
-		return nil, errors.New("mapped is not signed")
+		return nil, -1, errors.New("mapped is not signed")
 	}
 
 	tag := []byte(MozTag)
 	if len(mapped) < int(certTableOffset+certTableSize) {
-		return nil, fmt.Errorf("mapped is shorter than certTableOffset+certTableSize: %d", certTableOffset+certTableSize)
+		return nil, -1, fmt.Errorf("mapped is shorter than certTableOffset+certTableSize: %d", certTableOffset+certTableSize)
 	}
 	tagIndex := bytes.Index(mapped[certTableOffset:certTableOffset+certTableSize], tag)
 	if tagIndex == -1 {
-		return nil, errors.New("mapped does not contain dummy cert")
+		return nil, -1, errors.New("mapped does not contain dummy cert")
 	}
 
 	insertStart := int(certTableOffset) + tagIndex + len(tag)
 	if insertStart+len(code) >= len(mapped) {
-		return nil, errors.New("we are trying to write past the end of mapped")
+		return nil, -1, errors.New("we are trying to write past the end of mapped")
 	}
 
 	if insertStart+len(code) > int(certTableOffset+certTableSize) {
-		return nil, fmt.Errorf("code is longer than available cert table space")
+		return nil, -1, fmt.Errorf("code is longer than available cert table space")
 	}
 
 	modBytes := make([]byte, len(mapped))
+	nuls := make([]byte, MaxLength)
 	copy(modBytes, mapped)
-
+	copy(modBytes[insertStart:insertStart+MaxLength], nuls)
```

Comment:
My understanding is that `insertStart` is the position right after the moz tag (L66) and `MaxLength` is the length of the attribution code + the tag (L17). Aren't we writing `len(MozTag)` extra nul bytes here?

VBinDiff seems to confirm it (top window is an original installer, bottom is the modified installer):

<img width="568" alt="Screenshot 2023-02-22 at 09 47 30" src="https://user-images.githubusercontent.com/217628/220568831-b4fbd33c-fb3d-4624-99ec-86fae23579a9.png">

Reply:
Hm, yes, I think you're right. Whoops :smile:

Reply:
This makes me think that relying on the returned offset from WriteAttributionCode is maybe not a great idea. This may have been caught by the new test had it independently found the location...